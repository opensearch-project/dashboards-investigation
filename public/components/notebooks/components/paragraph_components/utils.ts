/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { NotebookType } from '../../../../../common/types/notebooks';

/**
 * Checks if this paragraph is in an agentic notebook and has been run before
 * @param notebookType - The type of notebook
 * @param paragraphIndex - The index of the current paragraph
 * @param totalParagraphs - The total number of paragraphs in the notebook
 * @returns true if the paragraph is agentic and has been run before (not the last paragraph)
 */
export const isAgenticRunBefore = (
  notebookType: NotebookType | undefined,
  paragraphIndex: number,
  totalParagraphs: number
): boolean => {
  return notebookType === NotebookType.AGENTIC && paragraphIndex < totalParagraphs - 1;
};
