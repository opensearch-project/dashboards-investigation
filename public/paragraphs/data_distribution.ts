/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ParagraphRegistryItem } from '../services/paragraph_service';
import { DataDistributionContainer } from '../components/notebooks/components/data_distribution/data_distribution_container';
import { AnomalyVisualizationAnalysisOutputResult } from '../../common/types/notebooks';

export const DataDistributionParagraphItem: ParagraphRegistryItem<AnomalyVisualizationAnalysisOutputResult> = {
  renderParagraph: DataDistributionContainer,
};
