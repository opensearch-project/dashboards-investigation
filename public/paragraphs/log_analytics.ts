/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { LogPatternAnalysisResult } from '../../common/types/log_pattern';
import { LogPatternContainer } from '../components/notebooks/components/log_analytics/log_pattern_container';
import { ParagraphRegistryItem } from '../services/paragraph_service';

export const LogPatternParagraphItem: ParagraphRegistryItem<LogPatternAnalysisResult> = {
  ParagraphComponent: LogPatternContainer,
};
