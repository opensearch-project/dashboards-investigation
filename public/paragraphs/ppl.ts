/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  getQueryOutputData,
  PPLParagraph,
  QueryObject,
} from '../components/notebooks/components/paragraph_components/ppl';
import { ParagraphRegistryItem } from '../services/paragraph_service';
import { callOpenSearchCluster } from '../plugin_helpers/plugin_proxy_call';
import { getClient } from '../services';

export const PPLParagraphItem: ParagraphRegistryItem<string, unknown, QueryObject> = {
  ParagraphComponent: PPLParagraph,
  getContext: async (paragraph) => {
    const { output, dataSourceMDSId } = paragraph || {};
    if (!output?.[0].result) {
      return '';
    }

    const query = output[0].result;
    const queryType = paragraph?.input.inputText.substring(0, 4) === '%sql' ? '_sql' : '_ppl';

    let queryObject = paragraph?.fullfilledOutput;

    if (!queryObject || queryObject.error) {
      queryObject = await callOpenSearchCluster({
        http: getClient(),
        dataSourceId: dataSourceMDSId,
        request: {
          path: `/_plugins/${queryType}`,
          method: 'POST',
          body: JSON.stringify({
            query,
          }),
        },
      });

      if (!queryObject || queryObject.error) {
        return '';
      }
    }

    const data = getQueryOutputData(queryObject);

    if (data.length === 0) {
      return '';
    }

    return `
          ## Step description
          This step executes PPL query and get response data for further research. Analyze these results as part of your investigation and consider how they relate to the overall issue. 
    
          ## Step result:
          User has executed the following PPL query: '${query}' which returned the following results:
          
          \`\`\`json
          ${JSON.stringify(data)}
          \`\`\`
        `;
  },
};
