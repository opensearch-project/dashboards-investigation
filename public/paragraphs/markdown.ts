/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { MarkdownParagraph } from '../components/notebooks/components/paragraph_components/markdown';
import { ParagraphRegistryItem } from '../services/paragraph_service';

export const MarkdownParagraphItem: ParagraphRegistryItem = {
  renderParagraph: MarkdownParagraph,
};
