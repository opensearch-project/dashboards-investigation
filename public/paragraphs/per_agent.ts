/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  DeepResearchInputParameters,
  DeepResearchOutputResult,
} from '../../common/types/notebooks';
import { DeepResearchParagraph } from '../components/notebooks/components/paragraph_components/deep_research';
import { ParagraphRegistryItem } from '../services/paragraph_service';

export const PERAgentParagraphItem: ParagraphRegistryItem<
  DeepResearchOutputResult | string,
  DeepResearchInputParameters
> = {
  renderParagraph: DeepResearchParagraph,
};
