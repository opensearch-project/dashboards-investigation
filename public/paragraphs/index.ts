/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AI_RESPONSE_TYPE,
  DASHBOARDS_VISUALIZATION_TYPE,
  DATA_DISTRIBUTION_PARAGRAPH_TYPE,
  DEEP_RESEARCH_PARAGRAPH_TYPE,
  LOG_PATTERN_PARAGRAPH_TYPE,
  OBSERVABILITY_VISUALIZATION_TYPE,
  OTHER_PARAGRAPH_TYPE,
} from '../../common/constants/notebooks';
import { DataDistributionParagraphItem } from './data_distribution';
import { LogPatternParagraphItem } from './log_analytics';
import { MarkdownParagraphItem } from './markdown';
import { OtherParagraphItem } from './other';
import { PERAgentParagraphItem } from './per_agent';
import { PPLParagraphItem } from './ppl';
import { VisualizationParagraphItem } from './visualization';

export const paragraphRegistry = [
  { types: [OTHER_PARAGRAPH_TYPE], item: OtherParagraphItem },
  { types: [DATA_DISTRIBUTION_PARAGRAPH_TYPE], item: DataDistributionParagraphItem },
  { types: [LOG_PATTERN_PARAGRAPH_TYPE], item: LogPatternParagraphItem },
  { types: [DEEP_RESEARCH_PARAGRAPH_TYPE, AI_RESPONSE_TYPE], item: PERAgentParagraphItem },
  {
    types: [
      DASHBOARDS_VISUALIZATION_TYPE.toUpperCase(),
      OBSERVABILITY_VISUALIZATION_TYPE.toUpperCase(),
      DASHBOARDS_VISUALIZATION_TYPE,
      OBSERVABILITY_VISUALIZATION_TYPE,
    ],
    item: VisualizationParagraphItem,
  },
  { types: ['ppl', 'sql'], item: PPLParagraphItem },
  { types: ['md'], item: MarkdownParagraphItem },
];
