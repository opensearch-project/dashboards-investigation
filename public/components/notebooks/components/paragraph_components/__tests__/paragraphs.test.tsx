/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { getOSDHttp } from '../../../../../../common/utils';
import { sampleParsedParagraghs1 } from '../../../../../../test/notebooks_constants';
import { Paragraphs } from '../paragraphs';

jest.mock('../../../../../../../../src/plugins/embeddable/public', () => ({
  ViewMode: {
    EDIT: 'edit',
    VIEW: 'view',
  },
}));

jest.mock('../../bubbleup/bubble_up_container', () => ({
  BubbleUpContainer: () => <div />,
}));

const mockFind = jest.fn().mockResolvedValue({
  savedObjects: [],
});

jest.mock('../../../../../framework/core_refs', () => ({
  coreRefs: {
    savedObjectsClient: {
      find: (options) => mockFind(options),
    },
  },
}));

describe('<Paragraphs /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the component', () => {
    const setPara = jest.fn();
    const paragraphSelector = jest.fn();
    const textValueEditor = jest.fn();
    const handleKeyPress = jest.fn();
    const addPara = jest.fn();
    const DashboardContainerByValueRenderer = jest.fn();
    const deleteVizualization = jest.fn();
    const setSelectedViewId = jest.fn();
    const deletePara = jest.fn();
    const runPara = jest.fn();
    const clonePara = jest.fn();
    const movePara = jest.fn();
    const para = sampleParsedParagraghs1[0];
    const utils = render(
      <Paragraphs
        ref={jest.fn()}
        para={para}
        setPara={setPara}
        dateModified="2023-11-01 01:02:03"
        index={0}
        paraCount={2}
        paragraphSelector={paragraphSelector}
        textValueEditor={textValueEditor}
        handleKeyPress={handleKeyPress}
        addPara={addPara}
        DashboardContainerByValueRenderer={DashboardContainerByValueRenderer}
        deleteVizualization={deleteVizualization}
        http={getOSDHttp()}
        selectedViewId="view_both"
        setSelectedViewId={setSelectedViewId}
        deletePara={deletePara}
        runPara={runPara}
        clonePara={clonePara}
        movePara={movePara}
        showQueryParagraphError={false}
        queryParagraphErrorMessage="error-message"
        dataSourceEnabled={false}
        paragraphs={[]}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();

    utils.getByLabelText('Open paragraph menu').click();
    utils.getByText('Run input').click();
    utils.getByLabelText('Open paragraph menu').click();
    utils.getByText('Duplicate').click();
    utils.getByLabelText('Open paragraph menu').click();
    utils.getByText('Delete').click();
  });

  it('use SavedObject find to fetch visualizations when dataSourceEnabled', () => {
    const setPara = jest.fn();
    const paragraphSelector = jest.fn();
    const textValueEditor = jest.fn();
    const handleKeyPress = jest.fn();
    const addPara = jest.fn();
    const DashboardContainerByValueRenderer = jest.fn();
    const deleteVizualization = jest.fn();
    const setSelectedViewId = jest.fn();
    const deletePara = jest.fn();
    const runPara = jest.fn();
    const clonePara = jest.fn();
    const movePara = jest.fn();
    const para = {
      uniqueId: 'paragraph_1a710988-ec19-4caa-83cc-38eb609427d1',
      isRunning: false,
      inQueue: false,
      showAddPara: false,
      isVizualisation: true,
      vizObjectInput: '{}',
      id: 1,
      inp: '# Type your input here',
      isOutputStale: false,
      paraDivRef: undefined,
      visEndTime: undefined,
      visSavedObjId: undefined,
      visStartTime: undefined,
      lang: 'text/x-md',
      editorLanguage: 'md',
      typeOut: ['MARKDOWN'],
      out: ['# Type your input here'],
    };
    const utils = render(
      <Paragraphs
        ref={jest.fn()}
        para={para}
        setPara={setPara}
        dateModified="2023-11-01 01:02:03"
        index={0}
        paraCount={2}
        paragraphSelector={paragraphSelector}
        textValueEditor={textValueEditor}
        handleKeyPress={handleKeyPress}
        addPara={addPara}
        DashboardContainerByValueRenderer={DashboardContainerByValueRenderer}
        deleteVizualization={deleteVizualization}
        http={getOSDHttp()}
        selectedViewId="view_both"
        setSelectedViewId={setSelectedViewId}
        deletePara={deletePara}
        runPara={runPara}
        clonePara={clonePara}
        movePara={movePara}
        showQueryParagraphError={false}
        queryParagraphErrorMessage="error-message"
        dataSourceEnabled={true}
        dataSourceManagement={{ ui: { DataSourceSelector: <></> } }}
        paragraphs={[]}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
    expect(mockFind).toHaveBeenCalledWith({
      type: 'visualization',
    });
  });
});
