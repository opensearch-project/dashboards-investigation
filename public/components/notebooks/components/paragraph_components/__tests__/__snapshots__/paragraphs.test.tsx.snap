// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paragraphs /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPanel--noBorder notebookParagraphWrapper"
>
  <div
    class="notebookHeaderActionMenu"
  >
    <div
      class="euiPopover euiPopover--anchorDownCenter"
    >
      <div
        class="euiPopover__anchor"
      >
        <button
          aria-label="Open paragraph menu"
          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--small"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="notebooks-paragraph notebooks-paragraph-light"
  >
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      style="width: 100%;"
    >
      <textarea
        class="euiTextArea euiTextArea--resizeVertical euiTextArea--fullWidth euiTextArea--compressed editorArea"
        data-test-subj="editorArea-"
        id="editorArea-"
        placeholder="Type %md on the first line to define the input type. 
Code block starts here."
        rows="3"
      >
        %md # Type your input here
      </textarea>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButton euiButton--primary euiButton--small"
          data-test-subj="runRefreshBtn-"
          type="button"
        >
          <span
            class="euiButtonContent euiButton__content"
          >
            <span
              class="euiButton__text"
            >
              Refresh
            </span>
          </span>
        </button>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      class="euiText euiText--small wrapAll markdown-output-text"
      data-test-subj="markdownOutputText"
    >
      <div
        class="markdown-body "
      />
    </div>
  </div>
</div>
`;

exports[`<Paragraphs /> spec use SavedObject find to fetch visualizations when dataSourceEnabled 1`] = `
<div
  class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPanel--noBorder notebookParagraphWrapper"
>
  <div
    class="notebookHeaderActionMenu"
  >
    <div
      class="euiPopover euiPopover--anchorDownCenter"
    >
      <div
        class="euiPopover__anchor"
      >
        <button
          aria-label="Open paragraph menu"
          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--small"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 6h4v4H0V6Zm1 1v2h2V7H1Zm5-1h4v4H6V6Zm1 1v2h2V7H7Zm5-1h4v4h-4V6Zm1 3h2V7h-2v2Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="notebooks-paragraph notebooks-paragraph-light"
  >
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
        style="min-width: 500px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
              id="random_html_id-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Title
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="euiComboBox euiComboBox--compressed"
                  role="combobox"
                >
                  <div
                    class="euiFormControlLayout euiFormControlLayout--compressed"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <div
                        class="euiComboBox__inputWrap euiComboBox__inputWrap--compressed euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                        data-test-subj="comboBoxInput"
                        tabindex="-1"
                      >
                        <p
                          class="euiComboBoxPlaceholder"
                        >
                          Find visualization
                        </p>
                        <div
                          class="euiComboBox__input"
                          style="font-size: 14px; display: inline-block;"
                        >
                          <input
                            aria-controls=""
                            data-test-subj="comboBoxSearchInput"
                            id="random_html_id"
                            role="textbox"
                            style="box-sizing: content-box; width: 2px;"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                      <div
                        class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                      >
                        <button
                          aria-label="Open list of options"
                          class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                          data-test-subj="comboBoxToggleListButton"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--small euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                            />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <button
              class="euiButton euiButton--primary euiButton--small"
              data-test-subj="para-input-visualization-browse-button"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Browse
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem"
      />
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
          id="random_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              Date range
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiSuperDatePicker__flexWrapper euiSuperDatePicker__flexWrapper--noUpdateButton"
            >
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiFormControlLayout euiFormControlLayout--compressed euiFormControlLayout--group euiSuperDatePicker"
                >
                  <div
                    class="euiPopover euiPopover--anchorDownLeft"
                  >
                    <div
                      class="euiPopover__anchor euiQuickSelectPopover__anchor"
                    >
                      <button
                        aria-label="Date quick select"
                        class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiFormControlLayout__prepend"
                        data-test-subj="superDatePickerToggleQuickMenuButton"
                        type="button"
                      >
                        <span
                          class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--small euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                            />
                          </svg>
                          <span
                            class="euiButtonEmpty__text euiQuickSelectPopover__buttonText"
                          >
                            <svg
                              aria-hidden="true"
                              class="euiIcon euiIcon--medium euiIcon-isLoading"
                              focusable="false"
                              height="16"
                              role="img"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                              />
                            </svg>
                          </span>
                        </span>
                      </button>
                    </div>
                  </div>
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <div
                      class="euiDatePickerRange euiDatePickerRange--inGroup"
                    >
                      <button
                        class="euiSuperDatePicker__prettyFormat euiSuperDatePicker__prettyFormat--compressed"
                        data-test-subj="superDatePickerShowDatesButton"
                      >
                        Last 15 minutes
                        <span
                          class="euiSuperDatePicker__prettyFormatLink"
                        >
                          Show dates
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButton euiButton--primary euiButton--small"
          data-test-subj="runRefreshBtn-"
          type="button"
        >
          <span
            class="euiButtonContent euiButton__content"
          >
            <span
              class="euiButton__text"
            >
              Refresh
            </span>
          </span>
        </button>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
  </div>
</div>
`;
